src/units/abilities.cpp

```diff
@@ -1094,133 +1091,33 @@ namespace { // Helpers for attack_type::special_active()
-    if(filter_student && filter_opponent && filter_attacker && filter_defender){
-	    return true;
-    }
-	return false;
-}
-
-/**
- * like to un->get_abilities() this returns the active abilities on this unit. That is. abilities of this unit and abilities of adjacent units that affect this unit. Unlike un->get_abilities() this  also checks for attack related filters.
- * @param[in]  un The current unit
- * @param[in]  ability The tagname of the ability, for filtered the abilities type
- * @param[in]  loc The location where we assume the un unit to be at
- * @param[in]  up The opponent
- * @param[in]  opp_loc The location where we assume the opponent to be at
- * @param[in]  attacker the condition for matches 'active_on=offense/defense' and @a special_affects_self/opponent
- * @param[in]  weapon, opp_weapon used for [filter_weapon] in @a ability_filter_fighter
- * @param[in]  affect_other if true we will only return abilities that affect the opponent otherwise we will return abilities that affect @a un
- *
- * @returns an unit_ability_list containing all abilities active in the current attack context.
- */
-static unit_ability_list list_leadership(const std::string& ability, unit_const_ptr un, unit_const_ptr up, const map_location& loc, const map_location& opp_loc, bool attacker, const_attack_ptr weapon, const_attack_ptr opp_weapon, bool affect_other)
-{
-	unit_ability_list abil = (*un).get_abilities(ability, loc);
-	for(unit_ability_list::iterator i = abil.begin(); i != abil.end();) {
-		bool abil_affect = false;//used for determine if abilities with special_affects_self or opponent must be erased of list
-		if(affect_other){
-			abil_affect = !special_affects_opponent((*i->first), attacker);
-		} else {
-			abil_affect = !special_affects_self((*i->first), attacker);
-		}
-		bool fighter_filter = !ability_apply_filter(un, up, ability, *i->first, loc, opp_loc, attacker, weapon, opp_weapon);
-		const std::string& active_on = (*i->first)["active_on"];
-		bool active_on_bool = !(active_on.empty() || (attacker && active_on == "offense") || (!attacker && active_on == "defense"));
-		if(active_on_bool || fighter_filter || abil_affect) {
-			i = abil.erase(i);
-		} else {
-			++i;
-		}
-	}
-	return abil;
-}
-
 unit_ability_list attack_type::list_ability(const std::string& ability) const
 {
-	const unit_map& units = display::get_singleton()->get_units();
-
-	unit_const_ptr self = self_;
-	unit_const_ptr other = other_;
-
-	if(self == nullptr) {
-		unit_map::const_iterator it = units.find(self_loc_);
-		if(it.valid()) {
-			self = it.get_shared_ptr().get();
-		}
-	}
-	if(other == nullptr) {
-		unit_map::const_iterator it = units.find(other_loc_);
-		if(it.valid()) {
-			other = it.get_shared_ptr().get();
-		}
-	}
-
-	// Make sure they're facing each other.
-	temporary_facing self_facing(self, self_loc_.get_relative_dir(other_loc_));
-	temporary_facing other_facing(other, other_loc_.get_relative_dir(self_loc_));
```

Comment:
maybe you can explain a little more why this it no longer needed?

Reply:
@gfgtdf this code exist already in special_active, now what this function is used in list_leadership, it is unecessary to conservate a duplicate. see https://github.com/wesnoth/wesnoth/blob/49f44914b243faebfc327b26df7bae84e23f0683/src/units/abilities.cpp#L1185